### PR TITLE
Reduce allocations each time a render pass begins in RenderingDevice.

### PR DESCRIPTION
**Pull request by DarioSamo**
Using https://github.com/godotengine/godot/pull/92206 to verify the result, allocations go down from ~1800 to ~1700 per frame in the TPS demo by making this fairly simple change to reuse the Vectors across function calls.

### CODE REVIEW

File: servers/rendering/rendering_device.cpp [hunk 1]
```diff
@@ -3332,12 +3332,14 @@ Error RenderingDevice::_draw_list_setup_framebuffer(Framebuffer *p_framebuffer,
...
 Error RenderingDevice::_draw_list_render_pass_begin(Framebuffer *p_framebuffer, InitialAction p_initial_color_action, FinalAction p_final_color_action, InitialAction p_initial_depth_action, FinalAction p_final_depth_action, const Vector<Color> &p_clear_colors, float p_clear_depth, uint32_t p_clear_stencil, Point2i p_viewport_offset, Point2i p_viewport_size, RDD::FramebufferID p_framebuffer_driver_id, RDD::RenderPassID p_render_pass) {
-	LocalVector<RDD::RenderPassClearValue> clear_values;
-	LocalVector<RDG::ResourceTracker *> resource_trackers;
-	LocalVector<RDG::ResourceUsage> resource_usages;
+	thread_local LocalVector<RDD::RenderPassClearValue> clear_values;
+	thread_local LocalVector<RDG::ResourceTracker *> resource_trackers;
+	thread_local LocalVector<RDG::ResourceUsage> resource_usages;
 	bool uses_color = false;
 	bool uses_depth = false;
 	clear_values.resize(p_framebuffer->texture_ids.size());
```
Comment (clayjohn) [inline servers/rendering/rendering_device.cpp:3340]: I'm a bit worried about clear_values not being cleared. Consider the case where `p_framebuffer->texture_ids.size()` is larger than the number of clear values. You could easily have entries in the clear_values array (at the end) that don't correspond to anything. Then, in `add_draw_list_begin` we treat the size of the clear_values array as the number of clear values. With the old code, it was guaranteed that those excess clear values were zeroed. But with this code, they may be non-zero. Accordingly, we should probably just clear clear_values to be safe

Comment (DarioSamo) [reply servers/rendering/rendering_device.cpp:3340]: Upon reviewing it... this one's a bit weird. The change to use the thread_local doesn't actually affect the behavior here compared to what was there before, but there's a lot of potential bugs that are visible now that we're looking at this code again.

> Consider the case where p_framebuffer->texture_ids.size() is larger than the number of clear values. You could easily have entries in the clear_values array (at the end) that don't correspond to anything.

This one's not a valid case and seems to be accounted for like this:
```
if (color_index < p_clear_colors.size()) {
				ERR_FAIL_INDEX_V(color_index, p_clear_colors.size(), ERR_BUG); // A bug.
```
So it'll fail to retrieve a clear color if the amount of texture IDs is bigger.

The one thing I notice that is very odd is that if this happens:
```
Texture *texture = texture_owner.get_or_null(p_framebuffer->texture_ids[i]);
if (!texture) {
	color_index++;
	continue;
}
```
Then the correspondence between the index in the clear values array and the texture ID is lost and the values will be shifted. But this seems to really depend on however the lower level APIs map these textures to the clear values. I'd require to make a more thorough review of that to make sure nothing will break if I do that change.

Changing it to a clear and only pushing the values that will get added seems like it'll potentially change the behavior and size can end up being different than it was before. I'm not sure I can recommend doing that unless we review that nothing else will break at the lower level if there's not enough values in the clear vector for the amount of texture IDs that were specified.

TL;DR: This PR in particular won't change the behavior at all, but it does seem like the code around it was sketchy before.

Comment (DarioSamo) [reply servers/rendering/rendering_device.cpp:3341]: I've added the extra clear to ensure there's pretty much no behavior difference compared to the current `master` other than reducing the allocations.

**Review by clayjohn — APPROVED** [review]
Looks good to me. This isn't a risky change, so we can sneak it in to 4.3. Fewer allocations is always better and this code gets called a lot

**Comment by akien-mga** [comment]
Thanks!
